pyomo/contrib/cp/sequence_var.py

```diff
@@ -0,0 +1,151 @@
+        if cls != SequenceVar:
+            return super(SequenceVar, cls).__new__(cls)
+        if args == ():
+            return ScalarSequenceVar.__new__(ScalarSequenceVar)
+        else:
+            return IndexedSequenceVar.__new__(IndexedSequenceVar)
+
+    def __init__(self, *args, **kwargs):
+        self._init_rule = Initializer(kwargs.pop('rule', None))
+        self._init_expr = kwargs.pop('expr', None)
+        kwargs.setdefault('ctype', SequenceVar)
+        super(SequenceVar, self).__init__(*args, **kwargs)
+
+        if self._init_expr is not None and self._init_rule is not None:
+            raise ValueError(
+                "Cannot specify both rule= and expr= for SequenceVar %s" % (self.name,)
+            )
+
+    def _getitem_when_not_present(self, index):
+        if index is None and not self.is_indexed():
+            obj = self._data[index] = self
+        else:
+            obj = self._data[index] = self._ComponentDataClass(component=self)
+        parent = self.parent_block()
+        obj._index = index
+
+        if self._init_rule is not None:
+            obj.set_value(self._init_rule(parent, index))
+        if self._init_expr is not None:
+            obj.set_value(self._init_expr)
+
+        return obj
+
+    def construct(self, data=None):
+        """
+        Construct the _SequenceVarData objects for this SequenceVar
+        """
+        if self._constructed:
+            return
+        self._constructed = True
+
+        if is_debug_set(logger):
+            logger.debug("Constructing SequenceVar %s" % self.name)
+
+        # Initialize index in case we hit the exception below
+        index = None
+        try:
+            if not self.is_indexed():
+                self._getitem_when_not_present(None)
+            if self._init_rule is not None:
+                for index in self.index_set():
+                    self._getitem_when_not_present(index)
+        except Exception:
+            err = sys.exc_info()[1]
+            logger.error(
+                "Rule failed when initializing sequence variable for "
+                "SequenceVar %s with index %s:\n%s: %s"
+                % (self.name, str(index), type(err).__name__, err)
+            )
+            raise
+
+    def _pprint(self):
+        """Print component information."""
+        headers = [
+            ("Size", len(self)),
+            ("Index", self._index_set if self.is_indexed() else None),
+        ]
+        return (
+            headers,
+            self._data.items(),
+            ("IntervalVars",),
+            lambda k, v: ['[' + ', '.join(iv.name for iv in v.interval_vars) + ']'],
+        )
+
+
+class ScalarSequenceVar(_SequenceVarData, SequenceVar):
+    def __init__(self, *args, **kwds):
+        _SequenceVarData.__init__(self, component=self)
+        SequenceVar.__init__(self, *args, **kwds)
+        self._index = UnindexedComponent_index
+
+
+class IndexedSequenceVar(SequenceVar):
+    pass
```

Comment:
Why bother with this if it's exactly the same as a `SequenceVar`?

Reply:
Ooooh, because I was thinking of overriding `__getitem__` to allow indexing by expressions, and then I thought better of it for the first pass. I guess I don't need it do I?

Reply:
The "standard" Pyomo object hierarchy makes (heavy) use of nearly/completely empty `Indexed*` component classes.  This is consistent with that model.

Reply:
Consistency is a good argument. Consider this resolved.